### ISSUE TITLE
Memory Leak in Queue

### ISSUE THREAD

**Issue by ppwwyyxx** [tensorflow/tensorflow#2942]
Version: nightly built for Python2+Linux+GPU about a week ago (perhaps earlier). Cannot test on latest binary due to #2939.

The following code:
```python
import tensorflow as tf
import numpy as np
import os

if __name__ == '__main__':
    input_vars = tf.placeholder(tf.float32, shape=(None, 60, 60, 4))
    queue = tf.FIFOQueue(1000, [tf.float32], name='queue')
    enqueue_op = queue.enqueue([input_vars])
    output = tf.reduce_mean(queue.dequeue())

    sess = tf.Session()
    sess.run(tf.initialize_all_variables())

    with sess.as_default():
        d = np.random.rand(16, 60, 60, 4)
        for k in range(990):
            enqueue_op.run(feed_dict={input_vars: d})
        cmd = "ps u " + str(os.getpid()) + " | tail -n 1 | awk '{print $6}'"
        while True:
            print "mem in KB: "
            os.system(cmd)
            for k in range(300):
                enqueue_op.run(feed_dict={input_vars: d})
                output.eval()
```
quickly consumes 10GB main memory.

Tested on ArchLinux as well as CentOS7.

**Comment by yaroslavvb** [comment]
See same thing on today head

**Comment by mrry** [comment]
After investigating this, it seems most likely that the cause is heap fragmentation, arising from the creation of a large number of NumPy arrays. The current `tf.Session` implementation is (inadvertantly) copying the incoming `feed_dict` values into new NumPy arrays on every step, and I've got a fix for that pending. This leads to a large amount of churn on the heap, and the default `malloc()` implementation doesn't appear to handle this well.

Once the fix is in, your example code should not leak, but we would recommend using [`tcmalloc`](http://goog-perftools.sourceforge.net/doc/tcmalloc.html) for more realistic programs. Running the same code with `tcmalloc` enabled shows no leak at all with unmodified TensorFlow 0.9. An alternative workaround is to set the `malloc()` options to be more aggressive about `mmap()`-ing for large allocations. For example, setting following the environment variable also eliminated the leak for me:

```
$ MALLOC_MMAP_THRESHOLD_=100000 python ...
```

**Comment by vrv** [comment]
Fixed by f68cbf1
